workers/price-reporter/src/manager/external_executor.rs

```diff
@@ -0,0 +1,467 @@
+    pub(crate) async fn execution_loop(mut self) -> Result<(), PriceReporterError> {
+        let mut job_receiver = self.job_receiver.take().unwrap();
+        let mut cancel_channel = self.cancel_channel.take().unwrap();
+
+        // Spawn WS handler loop, which forwards reads/writes over channels
+        let (msg_out_tx, mut msg_in_rx) = self.spawn_ws_handler_loop();
+
+        loop {
+            tokio::select! {
+                // Process price update from external price reporter
+                Some(price_message) = msg_in_rx.recv() => {
+                    self.handle_price_update(price_message).await.map_err(PriceReporterError::ExchangeConnectionError)?;
+                }
+
+                // Dequeue the next job from elsewhere in the local node
+                Some(job) = job_receiver.recv() => {
+                    if self.config.disabled {
+                        warn!("ExternalPriceReporter received job while disabled, ignoring...");
+                        continue;
+                    }
+
+                    tokio::spawn({
+                        let mut self_clone = self.clone();
+                        let msg_out_tx = msg_out_tx.clone();
+                        async move {
+                            if let Err(e) = self_clone.handle_job(job, msg_out_tx).await {
+                                error!("Error in ExternalPriceReporter execution loop: {e}");
+                            }
+                        }.instrument(info_span!("handle_job"))
+                    });
+                },
+
+                // Await cancellation by the coordinator
+                _ = cancel_channel.changed() => {
+                    info!("ExternalPriceReporter cancelled, shutting down...");
+                    return Err(PriceReporterError::Cancelled("received cancel signal".to_string()));
+                }
+            }
+        }
+    }
+
+    /// Spawns the task responsible for handling the websocket connection
+    /// with the external price reporter
+    fn spawn_ws_handler_loop(
+        &self,
+    ) -> (UnboundedSender<WebsocketMessage>, UnboundedReceiver<PriceMessage>) {
+        let price_reporter_url: Url = self
+            .config
+            .price_reporter_url
+            .clone()
+            .unwrap()
+            .parse()
+            .expect("Invalid price reporter URL");
+
+        let (msg_out_tx, msg_out_rx) = unbounded_channel();
+        let (msg_in_tx, msg_in_rx) = unbounded_channel();
+
+        let subscription_states = self.subscription_states.clone();
+
+        tokio::spawn(ws_handler_loop(
+            price_reporter_url,
+            subscription_states,
+            msg_out_rx,
+            msg_in_tx,
+        ));
+
+        (msg_out_tx, msg_in_rx)
+    }
+
+    /// Spawns the price streamer loop, which periodically publishes price
+    /// reports to the system bus for the given token pair
+    fn spawn_price_streamer_loop(self, base: Token, quote: Token) {
```

Comment:
Why not just push to the bus every time we receive a new price from the API? Simpler this way and doesn't require a separate loop? 

Ideally we get prices exactly on the cadence they're delivered from the exchange. We don't want to introduce extra latency via this loop's sleep. We can restructure the subscription states to hold entire price reports, the choice to make them atomic was a premature optimization. Though this second point isn't that important

Reply:
Now that I'm looking at the use below we wrap the atomic structure in a lock anyways so we don't save anything by having it atomic

Reply:
yeah fair point, i'll remove the price streaming loop. I'm not sure we can have the subscription states store price reports since we still need to store the latest price from each of the exchanges somewhere to calculate a price report for a pair. I'll make sure that we only acquire a write lock when adding a new token pair to the top-level mapping, so we should still get something out of atomics.